File: client/app/src/main/kotlin/com/paijan/pcpilot/background_worker/Receiver.kt
package com.paijan.pcpilot.background_worker

import com.paijan.pcpilot.utils.ClientMessage
import com.paijan.pcpilot.utils.Constants
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetSocketAddress
import java.net.SocketException
import java.util.concurrent.BlockingQueue

class Receiver(
        private val socket: DatagramSocket,
        private val receivedMessages: BlockingQueue<ClientMessage>,
        onThreadEnd: ThreadEndCallback
) : RunnableAdapter("Receiver", onThreadEnd) {
    private val packet: DatagramPacket = DatagramPacket(ByteArray(Constants.MAX_MESSAGE_SIZE), Constants.MAX_MESSAGE_SIZE)

    @Throws(InterruptedException::class)
    override fun runBody() {
        try {
            socket.receive(packet)
        } catch (e: SocketException) {
            throw InterruptedException()
        }

        for (message in ClientMessage.fromBytes(packet.data, packet.length, packet.socketAddress as InetSocketAddress)) {
            receivedMessages.add(message) // may throw
        }
    }
}

File: server/BackgroundWorker/Receiver.cpp
#include "Receiver.h"
#include "Utils/Constants.h"

void Receiver::onUpdate(BlockingQueue<ServerMessage> &receivedMessages, ConnectionlessSocket &socket) {
	if (!socket.receive(this->byteBuffer, this->addressBuffer, Constants::MAX_MESSAGE_SIZE, Constants::MANUAL_LOOP_BACK_RATE_MS)) {
		return;
	}

	ServerMessage::fromBytes(this->byteBuffer, this->addressBuffer, this->messageBuffer);
	for (auto &message : this->messageBuffer) {
		receivedMessages.push(std::move(message));
	}
}
